### PR TITLE
Dardar

### PR DESCRIPTION
**Pull request by adriaat**
- Module to read and resample dardar data with the same methods as for the Cloudsat DPC data
- Code to collocate with training and test data

### CODE REVIEW

**Comment by adriaat** [comment]
There is now a class to read and resample DARDAR data following the same methods as used for the data from Cloudsat DPC data. Even the same functions are imported and used, except for `subsample_iwc_and_height`, as the weights should be computed differently.

**Comment by adriaat** [comment]
The code to collocate with training and test data is a script in `scripts/collocate_dardar.py`.

Notes:

- The DARDAR data needs to be downloaded previously to the collocation (it's not done on-the-fly).

- Not all CloudSat DPC granules are foun in DARDAR data, implying that there will be less DARDAR collocations

- Few DARDAR pixels are missed in the collocation (see figure below). This can be fixed by extending the time bounds, e.g., with one second, but if we do that then it can mean that we actually get more pixels. Either way, it's only a few pixels and the geostationary data temporal resolution is much larger than the DARDAR temporal resolution, and I think it can be considered insignificant differences of 1 second.
  collocation_dardar_test
  Collocation example, black: 2C-ICE, blue: DARDAR. The differences are because of the time boundaries (playing with the time boundaries can make them match).

- The code is not intended to support the coarser CPCIR data prepared ("`cpcir2`").

- Some DARDAR profiles can have or consist only of NaN IWC values. An unfortunate resampling can happen if only one, full-NaN IWC profile falls in a bin, resulting in a "valid nan IWP". I adapted `ccic.data.write_scenes` to discard these profiles when writing the sparse file.

- The code can issue the following RunTimeWarning:
  ```
  Traceback (most recent call last):                                                                                                                                                   
    File "/home/amell/ccic/scripts/collocate_dardar.py", line 191, in <module>                                                                                                         
      resample_to_scene(f, GRIDSAT_GRID, dardar_files_dict, dst_gridsat)                                                                                                               
    File "/home/amell/ccic/scripts/collocate_dardar.py", line 103, in resample_to_scene                                                                                                
      ds_resampled = resample_data(                                                                                                                                                    
    File "/home/amell/ccic/ccic/data/cloudsat.py", line 512, in resample_data                                                                                                          
      cloudsat_file.add_retrieval_targets(                                                                                                                                             
    File "/home/amell/ccic/ccic/data/dardar.py", line 147, in add_retrieval_targets                                                                                                    
      iwc, height = subsample_iwc_and_height(iwc, source_height)                                                                                                                       
    File "/home/amell/ccic/ccic/data/dardar.py", line 41, in subsample_iwc_and_height                                                                                                  
      iwc = convolve(iwc, k, mode="valid", method="direct")                                                                                                                            
    File "/home/amell/miniconda3/envs/ccic/lib/python3.10/site-packages/scipy/signal/_signaltools.py", line 1430, in convolve                                                          
      return correlate(volume, _reverse_and_conj(kernel), mode, 'direct')                                                                                                              
    File "/home/amell/miniconda3/envs/ccic/lib/python3.10/site-packages/scipy/signal/_signaltools.py", line 262, in correlate                                                          
      z = _sigtools._correlateND(in1, in2, out, val)                                                                                                                                   
  RuntimeWarning: invalid value encountered in cast  
  ```
  but neither can I identify any consequence of this nor can I reproduce it. It seems to happen arbitrarily, and this is used before any random sampling is done.

File: ccic/data/dardar.py [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+                "liquid cloud",
+                "warm rain and liquid clouds",
+                "cold rain and liquid clouds",
+                "rain may be mixed with liquid",
+                "Multiple scattering due to supercooled water"
+            ),
+            "_FillValue": -1,
+        }
+
+    def get_iwp(self, dataset, above_ground=True):
```
Comment (simonpf) [inline ccic/data/dardar.py:236]: I would suggest moving member function, which don't use the self arguments out of the class to make them stand-alone functions on the module level.

It's more of a cosmetic change, but, at least in principle, it makes it easier to reuse the function.

File: ccic/data/dardar.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        if above_ground:
+            above_ground_mask = (self.get_surface_mask(dataset) == False)
+        else:
+            above_ground_mask = np.ones_like(iwc.shape, dtype=bool)
+
+        # Compute the integral, zeroing IWC if needed, and return
+        # Note that the height dimension is in decreasing order, hence the flip
+        return np.trapz((iwc * above_ground_mask)[..., ::-1], dataset.height[::-1], axis=1)
+
+    def get_surface_mask(self, dataset):
```
Comment (simonpf) [inline ccic/data/dardar.py:258]: See comment above.

**Review by simonpf — COMMENTED** [review]
This is great work. I just had two minor comments regarding code organization. And could you please run black on all the files? At least one line looked suspiciously long to me :)

File: ccic/data/dardar.py (hunk 2 above)
Comment (adriaat) [reply ccic/data/dardar.py]: Addressed with 771c8bf

File: ccic/data/dardar.py (hunk 1 above)
Comment (adriaat) [reply ccic/data/dardar.py]: Addressed with 771c8bf

**Comment by adriaat** [comment]
Thanks for the comments. Updated with your comments and with black formatting (except for one line, which black just makes it uggly; I kept that line identical to that in `ccic.data.cloudsat`). I also took the chance to handle an edge case which I came across after running the script for several hours.

**Pull request merged**